cpp/src/hash/fixed_hash_set.cuh

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utilities/device_atomics.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <hash/helper_functions.cuh>
+
+namespace cudf {
+
+/*
+ *  fixed set on a device
+ */
+template <typename Element, typename Hasher = default_hash<Element>,
+          typename Equality = equal_to<Element>>
+class device_fixed_hash_set {
+public:
+  device_fixed_hash_set(size_type hash_size, const size_type *hash_begin, const Element *hash_data)
+    : hash_size{hash_size}, hash_begin{hash_begin}, hash_data{hash_data}, hasher(), equals() {
+  }
+
+  bool __device__ contains(Element e) const {
+    size_type loc = hasher(e) % (2 * hash_size);
+
+    for (size_type i = hash_begin[loc] ; i < hash_begin[loc+1] ; ++i) {
+      if (equals(hash_data[i],e))
+        return true;
+    }
+
+    return false;
+  }
+
+private:
+  Hasher           hasher;
+  Equality         equals;
+  size_type        hash_size;
+  const size_type *hash_begin;
+  const Element   *hash_data;
+};
+
+/*
+ * Fixed size set on a device.  
+ *
+ */
+template <typename Element, typename Hasher = default_hash<Element>,
+          typename Equality = equal_to<Element>>
+class fixed_hash_set {
```

Comment:
Should be renamed because `set` implies ordered, unique values. 
```suggestion
class unordered_multiset{
```

Reply:
I have renamed it.  AFAIK, only C++ treats the term set as implying ordered elements.  Most other languages add an ordered set if you want that assumption.

Of course, this is C++ code... so I suppose we should follow the backwards way that the C++ standard interpreted this.